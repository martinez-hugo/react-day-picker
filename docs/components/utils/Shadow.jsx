// https://gist.github.com/janily/04d7fb0861e053d4679b38743ffc05a7

import { Component } from 'react';

import { createPortal } from 'react-dom';
import style from '!!url-loader!react-day-picker/dist/style.css';

export class Shadow extends Component {
  componentDidMount() {
    if (this.shadowAttached) return;
    this.shadowAttached = true;
    this.shadowRoot = this.node.attachShadow({ mode: this.props.mode });

    // Append a link to the parent element stylesheet
    const link = document.createElement('link');
    link.setAttribute('rel', 'stylesheet');
    link.setAttribute('href', style);
    this.shadowRoot.appendChild(link);

    this.forceUpdate();
  }

  render() {
    const { children, ...rest } = this.props;
    return (
      <div {...rest} ref={(node) => (this.node = node)}>
        {this.shadowRoot && createPortal(children, this.shadowRoot)}
      </div>
    );
  }
}
