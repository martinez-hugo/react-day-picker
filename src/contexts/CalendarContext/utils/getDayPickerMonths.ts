import {
  addDays,
  addWeeks,
  endOfISOWeek,
  endOfMonth,
  endOfWeek,
  getWeek,
  startOfISOWeek,
  startOfWeek
} from 'date-fns';

import {
  DayPickerDay,
  DayPickerMonth,
  DayPickerWeek
} from 'contexts/CalendarContext/DayPickerCalendar';

/** Return the {@link DayPickerMonth | DayPickerMonths} to display in the calendar. */
export function getDayPickerMonths(
  months: Date[],
  dates: Date[],
  options?: {
    fixedWeeks?: boolean | undefined;
    reverseMonths?: boolean | undefined;
    ISOWeek?: boolean | undefined;
    locale?: Locale | undefined;
    weekStartsOn?: 0 | 1 | 2 | 3 | 4 | 5 | 6 | undefined;
    firstWeekContainsDate?: 1 | 2 | 3 | 4 | 5 | 6 | 7 | undefined;
  }
) {
  const dayPickerMonths = months.reduce<DayPickerMonth[]>((months, month) => {
    const firstDateOfFirstWeek = options?.ISOWeek
      ? startOfISOWeek(month)
      : startOfWeek(month, options);
    const lastDateOfLastWeek = options?.ISOWeek
      ? endOfISOWeek(endOfMonth(month))
      : endOfWeek(endOfMonth(month), options);

    const monthDates = dates.filter((date) => {
      return date >= firstDateOfFirstWeek && date <= lastDateOfLastWeek;
    });

    if (options?.fixedWeeks && monthDates.length < 42) {
      const extraDates = dates.filter((date) => {
        return (
          date > lastDateOfLastWeek && date <= addDays(lastDateOfLastWeek, 7)
        );
      });
      monthDates.push(...extraDates);
    }

    const dayPickerWeeks = monthDates.reduce<DayPickerWeek[]>((weeks, date) => {
      const weekNumber = getWeek(date, options);
      const week = weeks.find((week) => week.weekNumber === weekNumber);

      const day = new DayPickerDay(date, month);
      if (!week) {
        weeks.push(new DayPickerWeek(weekNumber, [day]));
      } else {
        week.days.push(day);
      }
      return weeks;
    }, []);

    const dayPickerMonth = new DayPickerMonth(month, dayPickerWeeks);

    months.push(dayPickerMonth);
    return months;
  }, []);
  if (options?.reverseMonths) dayPickerMonths.reverse();

  return dayPickerMonths;
}
