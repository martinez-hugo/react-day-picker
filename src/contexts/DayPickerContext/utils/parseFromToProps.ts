import { endOfMonth, startOfDay, startOfMonth } from 'date-fns';
import { DayPickerBaseProps } from 'DayPicker';

/** Return the `fromDate` and `toDate` prop values values parsing the DayPicker props. */
export function parseFromToProps(
  props: Pick<
    DayPickerBaseProps,
    | 'fromYear'
    | 'toYear'
    | 'fromDate'
    | 'toDate'
    | 'fromMonth'
    | 'toMonth'
    | 'today'
  >
): { fromDate: Date | undefined; toDate: Date | undefined } {
  const { fromYear, toYear, fromMonth, toMonth, today = new Date() } = props;
  let { fromDate, toDate } = props;

  if (fromMonth) {
    fromDate = startOfMonth(fromMonth);
  } else if (fromYear) {
    fromDate = new Date(fromYear, 0, 1);
  }
  if (toMonth) {
    toDate = endOfMonth(toMonth);
  } else if (toYear) {
    toDate = new Date(toYear, 11, 31);
  }

  return {
    fromDate: startOfDay(fromDate ?? today),
    toDate: toDate ? startOfDay(toDate) : undefined
  };
}
