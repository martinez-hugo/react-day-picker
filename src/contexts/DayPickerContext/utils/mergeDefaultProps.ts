import { DayPickerProps, DaysSelectionMode } from 'DayPicker';

import { DayPickerContext } from 'contexts/DayPickerContext';

import { DataAttributes, defaultProps } from '../defaultProps';
import { parseFromToProps } from './parseFromToProps';

/** Merge the {@link defaultProps} with the props passed in. */
export function mergeDefaultProps<TMode extends DaysSelectionMode>(
  props: DayPickerProps,
  mode: DaysSelectionMode
) {
  const { fromDate, toDate } = parseFromToProps(props);

  const dataAttributes: DataAttributes = {};
  Object.entries(props).forEach(([key, val]) => {
    if (key.startsWith('data-')) {
      dataAttributes[key] = val;
    }
  });

  const value: DayPickerContext<TMode> = {
    ...defaultProps,
    ...props,
    mode,
    // captionLayout,
    classNames: {
      ...defaultProps.classNames,
      ...props.classNames
    },
    formatters: {
      ...defaultProps.formatters,
      ...props.formatters
    },
    fromDate,
    labels: {
      ...defaultProps.labels,
      ...props.labels
    },
    modifiers: {
      ...props.modifiers,
      ...defaultProps.modifiers,
      ...(props.selected ? { selected: props.selected } : {})
    },
    modifiersClassNames: {
      ...props.modifiersClassNames,
      ...defaultProps.modifiersClassNames
    },
    styles: {
      ...props.styles,
      ...defaultProps.styles
    },
    toDate,
    dataAttributes
  };

  return value;
}
