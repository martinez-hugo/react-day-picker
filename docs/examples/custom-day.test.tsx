import React from 'react';

import { renderExampleApp } from '../../test/renderExampleApp';
import { freezeBeforeAll } from '../../test/utils';
import Example from './custom-day';

freezeBeforeAll(new Date(2021, 10, 25));

test('should render time elements', () => {
  const { app } = renderExampleApp(<Example />);
  const timeElements = app.getElementsByTagName('time');
  expect(timeElements).toHaveLength(35);
});
