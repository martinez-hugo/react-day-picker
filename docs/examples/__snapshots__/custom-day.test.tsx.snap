// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render time elements 1`] = `
<div
  role="application"
>
  <div
    class="root"
  >
    <nav
      class="nav"
    >
      <button
        aria-label="Go to previous month"
        class="button_previous"
        name="previous-month"
      >
        <svg
          height="24px"
          viewBox="0 0 24 24"
          width="24px"
        >
          <polygon
            points="15 17.23 9.43 11.5 15 5.76 13.28 4 6 11.5 13.28 19"
          />
        </svg>
      </button>
      <button
        aria-label="Go to next month"
        class="button_next"
        name="next-month"
      >
        <svg
          height="24px"
          viewBox="0 0 24 24"
          width="24px"
        >
          <polygon
            points="9 17.23 14.56 11.5 9 5.76 10.71 4 18 11.5 10.71 19"
          />
        </svg>
      </button>
    </nav>
    <div
      class="months_wrapper"
    >
      <div
        class="month_grid_wrapper"
      >
        <div
          class="month_caption"
          id=":r0:"
          role="presentation"
        >
          November 2021
           
        </div>
        <div
          aria-label="November 2021"
          aria-multiselectable="false"
          class="month_grid"
          role="grid"
        >
          <div
            aria-rowindex="1"
            class="weekdays_row"
            role="row"
          >
            <span
              aria-colindex="1"
              aria-label="Sunday"
              class="weekday_columnheader"
              role="columnheader"
            >
              Su
            </span>
            <span
              aria-colindex="2"
              aria-label="Monday"
              class="weekday_columnheader"
              role="columnheader"
            >
              Mo
            </span>
            <span
              aria-colindex="3"
              aria-label="Tuesday"
              class="weekday_columnheader"
              role="columnheader"
            >
              Tu
            </span>
            <span
              aria-colindex="4"
              aria-label="Wednesday"
              class="weekday_columnheader"
              role="columnheader"
            >
              We
            </span>
            <span
              aria-colindex="5"
              aria-label="Thursday"
              class="weekday_columnheader"
              role="columnheader"
            >
              Th
            </span>
            <span
              aria-colindex="6"
              aria-label="Friday"
              class="weekday_columnheader"
              role="columnheader"
            >
              Fr
            </span>
            <span
              aria-colindex="7"
              aria-label="Saturday"
              class="weekday_columnheader"
              role="columnheader"
            >
              Sa
            </span>
          </div>
          <div
            class="month_rowgroup"
            role="rowgroup"
          >
            <div
              aria-rowindex="2"
              class="week_row"
              role="row"
            >
              <time
                aria-colindex="1"
                aria-hidden="true"
                class="day day_outside day_hidden"
                role="gridcell"
                tabindex="-1"
              >
                31
              </time>
              <time
                aria-colindex="2"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                1
                <div>
                  (first day)
                </div>
              </time>
              <time
                aria-colindex="3"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                2
              </time>
              <time
                aria-colindex="4"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                3
              </time>
              <time
                aria-colindex="5"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                4
              </time>
              <time
                aria-colindex="6"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                5
              </time>
              <time
                aria-colindex="7"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                6
              </time>
            </div>
            <div
              aria-rowindex="3"
              class="week_row"
              role="row"
            >
              <time
                aria-colindex="1"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                7
              </time>
              <time
                aria-colindex="2"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                8
              </time>
              <time
                aria-colindex="3"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                9
              </time>
              <time
                aria-colindex="4"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                10
              </time>
              <time
                aria-colindex="5"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                11
              </time>
              <time
                aria-colindex="6"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                12
              </time>
              <time
                aria-colindex="7"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                13
              </time>
            </div>
            <div
              aria-rowindex="4"
              class="week_row"
              role="row"
            >
              <time
                aria-colindex="1"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                14
              </time>
              <time
                aria-colindex="2"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                15
              </time>
              <time
                aria-colindex="3"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                16
              </time>
              <time
                aria-colindex="4"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                17
              </time>
              <time
                aria-colindex="5"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                18
              </time>
              <time
                aria-colindex="6"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                19
              </time>
              <time
                aria-colindex="7"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                20
              </time>
            </div>
            <div
              aria-rowindex="5"
              class="week_row"
              role="row"
            >
              <time
                aria-colindex="1"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                21
              </time>
              <time
                aria-colindex="2"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                22
              </time>
              <time
                aria-colindex="3"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                23
              </time>
              <time
                aria-colindex="4"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                24
              </time>
              <time
                aria-colindex="5"
                class="day day_today"
                role="gridcell"
                tabindex="-1"
              >
                25
              </time>
              <time
                aria-colindex="6"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                26
              </time>
              <time
                aria-colindex="7"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                27
              </time>
            </div>
            <div
              aria-rowindex="6"
              class="week_row"
              role="row"
            >
              <time
                aria-colindex="1"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                28
              </time>
              <time
                aria-colindex="2"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                29
              </time>
              <time
                aria-colindex="3"
                class="day"
                role="gridcell"
                tabindex="-1"
              >
                30
              </time>
              <time
                aria-colindex="4"
                aria-hidden="true"
                class="day day_outside day_hidden"
                role="gridcell"
                tabindex="-1"
              >
                1
              </time>
              <time
                aria-colindex="5"
                aria-hidden="true"
                class="day day_outside day_hidden"
                role="gridcell"
                tabindex="-1"
              >
                2
              </time>
              <time
                aria-colindex="6"
                aria-hidden="true"
                class="day day_outside day_hidden"
                role="gridcell"
                tabindex="-1"
              >
                3
              </time>
              <time
                aria-colindex="7"
                aria-hidden="true"
                class="day day_outside day_hidden"
                role="gridcell"
                tabindex="-1"
              >
                4
              </time>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
